### PR TITLE
Grid Support

### LINKED ISSUE #176: Support grid
It needs a new module.

https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Grids

https://css-tricks.com/snippets/css/complete-guide-grid/

### PR DESCRIPTION
**Pull request by CharlesTaylor7**
Resolves #176 

Explicit support for all grid properties listed on mdn, except for a handful of shorthand properties:
 - [grid](https://developer.mozilla.org/en-US/docs/Web/CSS/grid)
 - [grid-row]( https://developer.mozilla.org/en-US/docs/Web/CSS/grid-row)
 - [grid-column]( https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column)
 - [grid-template](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template)

The above shorthands have involved syntax which would make them a pain to support. The finer grained methods still encapsulate all the grid layout options.

I'm mostly happy with how this pr turned out. The main shortcoming for me is `upcast`. I tried a number of other things, but none of them worked very well.  What would be nice here is Impredicative Polymorphism. I want to be able to create a list of type `[forall a. Size a]` for `gridTemplateRows`/`gridTemplateColumns`, but GHC won't oblige me. Maybe something to revisit when we get [Quick Look](https://gitlab.haskell.org/ghc/ghc/-/issues/18126)?

### CODE REVIEW

**Comment by turion** [comment]
Exciting :) do assign me as a reviewer when you're ready.

**Comment by CharlesTaylor7** [comment]
@turion Ready for review. I don't think have permissions for explicitly assigning you as reviewer.

File: spec/Clay/SizeSpec.hs [hunk 1]
```diff
@@ -47,8 +42,6 @@ spec = do
       sizeRepr (em 2 @+@ px 1) `shouldBe` "calc(2em + 1px)"
     it "returns calc for nested sum" $
       sizeRepr (em 2 @+@ pt 1 @+@ px 3) `shouldBe` "calc((2em + 1pt) + 3px)"
-    it "returns prefixed calc for simple sum" $
```
Comment (turion) [inline spec/Clay/SizeSpec.hs:50]: Why is this removed?

File: spec/Common.hs [hunk 2]
```diff
@@ -1,8 +1,16 @@
-module Common where
+module Common
+  ( shouldRenderFrom
```
Comment (turion) [inline spec/Common.hs:2]: I think there is no need to filter the exports. Or was there a name clash?

File: src/Clay/Grid.hs [hunk 3]
```diff
@@ -1,58 +1,271 @@
...
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocation_Data . GridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
+  = GridLocation_Keyword Value
```
Comment (turion) [inline src/Clay/Grid.hs:166]: Please remove the underscore.

File: src/Clay/Grid.hs [hunk 4]
```diff
@@ -1,58 +1,271 @@
...
+  | GridLocation_Data GridLocationData
```
Comment (turion) [inline src/Clay/Grid.hs:167]: Let us try and rethink the naming here. Having nearly the same constructor names and type names is a source of confusion.

File: src/Clay/Grid.hs [hunk 5]
```diff
@@ -1,58 +1,271 @@
...
+instance Val GridLocationData where
+  value (GridLocationData isSpan indexAndOrGridArea) =
+    if isSpan == Span
+    then value (span :: Value, indexAndOrGridArea)
+    else value indexAndOrGridArea
+
+pattern GridIndex :: Integer -> GridLocation
+pattern GridIndex n = GridLocation_Data (GridLocationData NoSpan (This n))
+
+instance Num GridLocation where
```
Comment (turion) [inline src/Clay/Grid.hs:192]: There should not be a `Num` instance if most of the methods are `error`. Rather give the two methods you're interested in separate functions.

File: src/Clay/Grid.hs [hunk 6]
```diff
@@ -1,58 +1,271 @@
...
+-- | Property defines the template for grid layout
+gridTemplateAreas :: GridTemplateAreas -> Css
+gridTemplateAreas = key "grid-template-areas"
+
+newtype GridTemplateAreas = GridTemplateAreas Value
+  deriving (Val, None, Inherit, Initial, Unset)
+
+instance IsList GridTemplateAreas where
+  type Item GridTemplateAreas = [GridArea]
+  toList = error "toList GridTemplateAreas is not defined"
```
Comment (turion) [inline src/Clay/Grid.hs:215]: Like for the `Num` instance, you probably just want to be able to write list literals. But that leads to a partial interface, and it is confusing if at some point people try to probe it further and discover this error.

File: src/Clay/Grid.hs [hunk 7]
```diff
@@ -1,58 +1,271 @@
...
+      rows = coerce areas :: [[Text]]
+      wrapInParens text = "\"" <> text <> "\""
+      convertRow = wrapInParens . Text.intercalate " "
+    in
+      value $
+      Text.intercalate "\n" $
+      fmap convertRow $
+      rows
+
+-- | toList will throw when your grid template areas are invalid
```
Comment (turion) [inline src/Clay/Grid.hs:236]: Here and elsewhere: Use ticks (') for haddock references.

File: src/Clay/Grid.hs [hunk 8]
```diff
@@ -1,58 +1,271 @@
...
+instance IsList GridTemplateAreas where
+  type Item GridTemplateAreas = [GridArea]
+  toList = error "toList GridTemplateAreas is not defined"
+  fromList = GridTemplateAreas . value . fromList'
+    where
+      fromList' :: [Item GridTemplateNamedAreas] -> GridTemplateNamedAreas
+      fromList' = fromList
+
+-- have to create a newtype to override the Val instance for lists
+newtype GridTemplateNamedAreas = GridTemplateNamedAreas { unGridTemplateNamedAreas :: [[GridArea]] }
```
Comment (turion) [inline src/Clay/Grid.hs:222]: In this particular case, `toList` should be definable. Even derivable: Can't you just use https://hackage.haskell.org/package/derive-monoid and derive `IsList`?

File: src/Clay/Grid.hs [hunk 9]
```diff
@@ -1,58 +1,271 @@
...
+      value $
+      Text.intercalate "\n" $
+      fmap convertRow $
+      rows
+
+-- | toList will throw when your grid template areas are invalid
+instance IsList GridTemplateNamedAreas where
+  type Item GridTemplateNamedAreas = [GridArea]
+  toList = unGridTemplateNamedAreas . coerce
+  fromList = fromRightOrThrow . mkGridTemplateNamedAreas
```
Comment (turion) [inline src/Clay/Grid.hs:240]:
```suggestion
  fromList = either throw id . mkGridTemplateNamedAreas
```

File: src/Clay/Grid.hs [hunk 10]
```diff
@@ -1,58 +1,271 @@
...
+    in
+      value $
+      Text.intercalate "\n" $
+      fmap convertRow $
+      rows
+
+-- | toList will throw when your grid template areas are invalid
+instance IsList GridTemplateNamedAreas where
+  type Item GridTemplateNamedAreas = [GridArea]
+  toList = unGridTemplateNamedAreas . coerce
```
Comment (turion) [inline src/Clay/Grid.hs:239]: Why do you need `coerce`?

File: src/Clay/Grid.hs [hunk 11]
```diff
@@ -1,58 +1,271 @@
...
+      Left GridTemplateNamedAreas_EmptyRow
+
+    when (any (/= longest) counts)  $
+      Left GridTemplateNamedAreas_NotRectangular
+
+    Right $ GridTemplateNamedAreas rows
+
+-- | Failure modes for the smart constructor
+data InvalidGridTemplateNamedAreas
+  = GridTemplateNamedAreas_Empty
```
Comment (turion) [inline src/Clay/Grid.hs:266]: Remove the underscores.

File: src/Clay/Grid.hs [hunk 12]
```diff
@@ -1,58 +1,271 @@
...
+      fromRightOrThrow (Left e) = throw e
+
+-- | Smart constructor for GridTemplateNamedAreas
+mkGridTemplateNamedAreas :: [[GridArea]] -> Either InvalidGridTemplateNamedAreas GridTemplateNamedAreas
+mkGridTemplateNamedAreas rows = do
+    let
+      counts = fmap length (coerce rows :: [[GridArea]])
+      longest = maximum counts
+
+    when (null rows ) $
```
Comment (turion) [inline src/Clay/Grid.hs:253]:
```suggestion
    when (null rows) $
```

File: src/Clay/Grid.hs (hunk 9 above)
Comment (turion) [inline src/Clay/Grid.hs:240]: Is the exception thrown at compile time or at runtime?

File: clay.cabal [hunk 13]
```diff
@@ -75,7 +75,8 @@ Library
   Build-Depends:
     base  >= 4.11 && < 5,
     mtl   >= 1    && < 2.3,
-    text  >= 0.11 && < 1.3
+    text  >= 0.11 && < 1.3,
+    these < 1.2
```
[the remark this answers is not in the document]
Comment (CharlesTaylor7) [reply clay.cabal:79]: I looked at the earliest version of the package on hackage, and it has the constructurs I'm making use of.
But I can add a lower bound for completeness.

File: src/Clay/Grid.hs (hunk 9 above)
Comment (CharlesTaylor7) [reply src/Clay/Grid.hs:240]: Runtime

File: spec/Common.hs (hunk 2 above)
Comment (CharlesTaylor7) [reply spec/Common.hs:2]: No name clash. I introduced a new top level binding of `testRender` & was preventing its export. It's not a big deal either way.

File: src/Clay/Grid.hs (hunk 10 above)
Comment (CharlesTaylor7) [reply src/Clay/Grid.hs:239]: I'm not sure why that's here. Looks like a relic.

File: src/Clay/Grid.hs (hunk 8 above)
Comment (CharlesTaylor7) [reply src/Clay/Grid.hs:222]: `toList` is defined for this. 
I could derive IsList automatically, but then I wouldn't validate the shape of the grid template.
If you generate a grid a non rectangular grid template then the browser will silently ignore that piece of css.
Crashing at generation of css seemed like a better experience.

I did consider using type level finite vectors to check the shape at compile time, but opted for runtime validation to keep error message simple.

**Comment by CharlesTaylor7** [comment]
@turion To be clear, you're very opposed to partial `Num` instances which exist just for number literals?

But are you open to partial `IsList` instances or no?
I will point out that `NonEmpty` has a partial `fromList` implementation for its `IsList` instance. I find it useful, but I recognize that its debatable and a matter of taste.

File: src/Clay/Size.hs [hunk 14]
```diff
@@ -88,12 +91,19 @@ data LengthUnit
 data Percentage
 
 -- | When combining percentages with units using calc, we get a combination
-data Combination
+-- | Any unit or combination of units
+data AnyUnit
+
+-- | Upcast a size unit
+upcast :: Size a -> Size AnyUnit
+upcast = coerce
```
[the remark this answers is not in the document]
Comment (CharlesTaylor7) [reply src/Clay/Size.hs:99]: Can you elaborate?
What's the functional difference from the user perspective?
It's more work to embed it in a constructor, I now have to add cases to `sizeToText` & the `Val` instance that do nothing but unwrap.

[the remark this answers is not in the document]
Comment (turion) [reply src/Clay/Size.hs:99]: From the user perspective it's mostly the same I hope. But from the maintainer's perspective a huge burden is gone because noone has to worry about `coerce` breaking anything.

File: spec/Common.hs (hunk 2 above)
Comment (turion) [reply spec/Common.hs:2]: Ah, I'm fine if you don't restrict the exports in that case.

File: clay.cabal (hunk 13 above)
[the remark this answers is not in the document]
Comment (turion) [reply clay.cabal:79]: Yes that would be nice, then noone is going to ask themselves that question anymore.

File: src/Clay/Grid.hs (hunk 6 above)
Comment (turion) [reply src/Clay/Grid.hs:215]: Also in this case, can't you just write the definition of `toList`?

File: src/Clay/Grid.hs (hunk 5 above)
Comment (turion) [reply src/Clay/Grid.hs:192]: In this particular case, it seems that a `Num` instance is confusing because many values for `GridLocation` aren't numbers. Right?

**Comment by turion** [comment]
> @turion To be clear, you're very opposed to partial `Num` instances which exist just for number literals?
> 
> But are you open to partial `IsList` instances or no?
> I will point out that `NonEmpty` has a partial `fromList` implementation for its `IsList` instance. I find it useful, but I recognize that its debatable and a matter of taste.

In my own libraries, I would not allow any partial functions that are reachable for library users. In `clay`, which I didn't author and only maintain so it is preserved for the community, there already are a lot of partial functions. They partly exist in places where I don't see their justification: For example `Size Percentage` has a `Num` instance without a `(+)` function, but with a refactoring of the code it would be possible to define `(+)` and all the other operations. Given enough time and dev effort, we can of course remove all of these warts, but time and effort are limited resources.

In particular for a CSS processor many people will take the stance that in a sense clay is always run "before runtime" in that CSS is generated before the webserver starts. Still these errors don't show up during type checking and are therefore not in the typical Haskell dev workflow. You'd think that your program will run when it compiles, but it doesn't necessarily here.

Since I don't want to undermine your efforts, I'd be willing to accept these partial functions if you really think they add value and cannot made be total with reasonable effort.

But if it's just for being able to write an integer literal I'm not so sure it's all that much added value. Consider `shadow 23 42` vs `shadow (px 23) (pct 42)`. The first expression is weirdly polymorphic in the size type in that you can later decide whether your 23 is percentage or pixels. But the idiomatic style is the second one anyways, which specifies the length units and types. It's clearer and not very long. The `Num` instance wasn't even used. I'd expect it to be similar in your applications, but of course you will be able to judge this better.

File: src/Clay/Grid.hs (hunk 8 above)
Comment (srid) [reply src/Clay/Grid.hs:222]: > Crashing at generation of css seemed like a better experience.

Crashing at runtime is generally a bad idea for a Haskell library. Clay is not necessarily used only to generate CSS, but also used in Haskell webapps to generate it in *real-time*.

Comment (turion) [reply src/Clay/Grid.hs:222]: @srid I tend to agree. Although it will typically "reliably crash" on startup, so you'll probably notice on deployment. But for Haskell, which I associate with "if it compiles it works", this is also not acceptable in my opinion. Maybe in the long run, Template Haskell should be used for such cases, and then the crash is again on compile time.

**Comment by ddssff** [comment]
Would love to see this merged.

**Comment by turion** [comment]
Yes, that would be great :)

@CharlesTaylor7 do you want to continue? I hope my review so far wasn't to rigorous to demotivate. Reading back what I wrote it sounds quite harsh in my ears. If it did that for you as well, I'm really sorry, and I'll try to write in a friendlier tone :)

If you have any questions on how to proceed, feel free to ask.

**Comment by CharlesTaylor7** [comment]
> Yes, that would be great :)
> 
> @CharlesTaylor7 do you want to continue? I hope my review so far wasn't to rigorous to demotivate. Reading back what I wrote it sounds quite harsh in my ears. If it did that for you as well, I'm really sorry, and I'll try to write in a friendlier tone :)
> 
> If you have any questions on how to proceed, feel free to ask.

@turion I would like to complete this PR. I hate loose ends and do appreciate quality focused, thorough code reviews. I have no issue with you, just lost interest with this last year. I have the time and motivation to tackle this now.

File: src/Clay/Grid.hs (hunk 6 above)
Comment (CharlesTaylor7) [reply src/Clay/Grid.hs:217]: No, in this case you can't write `toList`, only `fromList`. GridTemplateAreas is defined as a newtype around `Value`. Some but not all values are lists. So you can convert a list to a value, but not vice versa. 

That said, I'm stepping through this and going to remove the partial type class instances. 
I'm going to replace them with pattern synonyms  or smart constructors.

**Comment by CharlesTaylor7** [comment]
@turion https://github.com/sebastiaanvisser/clay/pull/210#discussion_r463956124
Do you have a recommended naming scheme for handling complex data definitions?
I think it's a benefit, not a disadvantage, for the record type name, to have a similar name to the sum type constructor's name, this links the two clearly. Though I do agree there needs to be a convention to know which is which.

For sum types, I took the approach of prefixing constructor names with the name of type. This can be done with or without underscores. I removed the underscores, because they are at odds with the rest of the library which doesn't use underscores. (Though, I do think the underscores make it easier to visually parse).

File: src/Clay/Size.hs (hunk 14 above)
[the remark this answers is not in the document]
Comment (turion) [reply src/Clay/Size.hs:99]: I have to admit that I completely misunderstood what you did here. I mistook `coerce` for `unsafeCoerce`, which is - as the name says - unsafe, whereas `coerce` is completely fine. It will never break anything because the type checker ensures that coercion only happens when the internal representations are indeed the same. If that was your goal, I'm sorry about the misdirection: feel free to disregard the stuff I said about GADTs, and let's make everyone's life easier by using `coerce`.

File: src/Clay/Grid.hs [hunk 15]
```diff
@@ -1,58 +1,249 @@
...
+instance Auto    GridLocation where auto    = GridLocationKeyword auto
+instance Inherit GridLocation where inherit = GridLocationKeyword inherit
+instance Initial GridLocation where initial = GridLocationKeyword initial
+instance Unset   GridLocation where unset   = GridLocationKeyword unset
+
+-- See under syntax:
+-- https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
+-- either grid index and/or named grid area is required, but span is optional
+-- Note: constuctor is not exported, 'GridIndex, or 'gridLocationData to fill this info in
+data GridLocationData = MkGridLocationData IsSpan (These Integer GridArea)
```
Comment (turion) [inline src/Clay/Grid.hs:181]: The Constructor should have the same name as the type.

File: src/Clay/Grid.hs [hunk 16]
```diff
@@ -1,58 +1,249 @@
...
+
+instance Auto    GridLocation where auto    = GridLocationKeyword auto
+instance Inherit GridLocation where inherit = GridLocationKeyword inherit
+instance Initial GridLocation where initial = GridLocationKeyword initial
+instance Unset   GridLocation where unset   = GridLocationKeyword unset
+
+-- See under syntax:
+-- https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
+-- either grid index and/or named grid area is required, but span is optional
+-- Note: constuctor is not exported, 'GridIndex, or 'gridLocationData to fill this info in
```
Comment (turion) [inline src/Clay/Grid.hs:180]:
```suggestion
-- Note: constructor is not exported, 'GridIndex, or 'gridLocationData to fill this info in
```

**Comment by turion** [comment]
> I would like to complete this PR.

Fantastic, looking forward to reviewing further :)

> @turion [#210 (comment)](https://github.com/sebastiaanvisser/clay/pull/210#discussion_r463956124) Do you have a recommended naming scheme for handling complex data definitions? I think it's a benefit, not a disadvantage, for the record type name, to have a similar name to the sum type constructor's name, this links the two clearly. Though I do agree there needs to be a convention to know which is which.
> 
> For sum types, I took the approach of prefixing constructor names with the name of type. This can be done with or without underscores. I removed the underscores, because they are at odds with the rest of the library which doesn't use underscores. (Though, I do think the underscores make it easier to visually parse).

This is a tough problem that I also often have. A few rules of thumb that I have for these situations:

* Newtypes & product types have to have the same constructor name as the type name (`GridLocationData` in this example)
* There cannot be a constructor with the name of an existing, distinct type (because otherwise people will think that the previous rule applies)
* In sum types, ideally the constructor name doesn't need to repeat the type name. If collisions occur, avoid them by first thinking about different naming, and then putting types into submodules.

Specifically for `clay`, I'd add another rule of thumb:

* If some Haskell definition matches exactly some known CSS concept, the naming should exactly mirror the CSS naming (adapted to camelCase and UpperCamelCase).

In your case though, I'd simply remove the `GridLocationData` type and add the fields to the corresponding constructor of the `GridLocation` type, though. The `GridLocationData` type isn't used anywhere else, and the total size of instance code will not be higher I think.

File: src/Clay/Grid.hs [hunk 17]
```diff
@@ -1,58 +1,249 @@
...
+gridColumnEnd :: GridLocation -> Css
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocationData . MkGridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
```
Comment (turion) [inline src/Clay/Grid.hs:164]: Since this is not a known CSS concept (as far as I know), describe a bit what it is and what it is good for.

File: src/Clay/Grid.hs [hunk 18]
```diff
@@ -1,58 +1,249 @@
...
+
+blankGridArea :: GridArea
+blankGridArea = GridArea "."
+
+newtype GridArea = GridArea Text
+  deriving (IsString, Val)
+
+-------------------------------------------------------------------------------
+
+gridRowStart :: GridLocation -> Css
```
Comment (turion) [inline src/Clay/Grid.hs:146]: But I thought grid rows don't support spans? See https://www.w3schools.com/cssref/pr_grid-row-start.asp. Maybe `GridLocation` should be `GridColumnLocation`?

File: src/Clay/Common.hs [hunk 19]
```diff
@@ -15,17 +15,21 @@ import Data.String (IsString)
...
+class Baseline   a where baseline   :: a
+class Center     a where center     :: a
+class Inherit    a where inherit    :: a
+class None       a where none       :: a
+class Normal     a where normal     :: a
+class Visible    a where visible    :: a
+class Hidden     a where hidden     :: a
+class Initial    a where initial    :: a
+class Unset      a where unset      :: a
+class MinContent a where minContent :: a
```
Comment (turion) [inline src/Clay/Common.hs:29]: Is there any chance that some other types than `Value` will also be instances of your new classes?

File: src/Clay/Grid.hs [hunk 20]
```diff
@@ -1,58 +1,249 @@
...
+  , ExistentialQuantification
+  , PatternSynonyms
+  #-}
+-- | Implementation of <https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout>.
 module Clay.Grid
-( gridGap
-, gridTemplateColumns
-)
-where
+  ( gap
```
Comment (turion) [inline src/Clay/Grid.hs:11]: Are there many definitions in this module that are _not_ exported? If everything is exported, we can also simply not restrict the exports.

File: src/Clay/Grid.hs [hunk 21]
```diff
@@ -1,58 +1,249 @@
...
+import Data.These
+import GHC.Exts (IsList(..))
+import Control.Exception (Exception(..), throw)
+import Control.Monad (when)
+
 
 -- | Property sets the gaps (gutters) between rows and columns.
-gridGap :: Size a -> Css
-gridGap = key "grid-gap"
+-- Sets both "gap" & "grid-gap"
```
Comment (turion) [inline src/Clay/Grid.hs:65]:
```suggestion
--
-- Sets both "gap" & "grid-gap"
```

File: src/Clay/Grid.hs [hunk 22]
```diff
@@ -1,58 +1,249 @@
...
+-- to quote: https://developer.mozilla.org/en-US/docs/Web/CSS/gap
+-- "CSS Grid Layout initially defined the grid-gap property. This prefixed property is being replaced by gap. However, in order to support browsers that implemented grid-gap and not gap for grid, you will need to use the prefixed property"
+gap :: Size a -> Css
+gap = key "gap" <> key "grid-gap"
+
+-- | Property sets the size of the gap (gutter) between an element's grid rows.
+-- Sets both "row-gap" & "grid-row-gap"
```
Comment (turion) [inline src/Clay/Grid.hs:72]:
```suggestion
--
-- Sets both "row-gap" & "grid-row-gap"
```

At least I believe this will make a paragraph in haddock.

File: src/Clay/Grid.hs [hunk 23]
```diff
@@ -1,58 +1,249 @@
...
+-- to quote: https://developer.mozilla.org/en-US/docs/Web/CSS/row-gap
+-- "CSS Grid Layout initially defined the grid-row-gap property. This prefixed property is being replaced by row-gap. However, in order to support browsers that implemented grid-row-gap and not row-gap for grid, you will need to use the prefixed property."
+rowGap :: Size a -> Css
+rowGap = key "row-gap" <> key "grid-row-gap"
+
+-- | Property sets the size of the gap (gutter) between an element's grid columns.
+-- Sets both "column-gap" & "grid-column-gap"
```
Comment (turion) [inline src/Clay/Grid.hs:79]: Same here

File: src/Clay/Grid.hs [hunk 24]
```diff
@@ -1,58 +1,249 @@
...
+-- | Property defines the line names and track sizing functions of the grid rows.
+gridTemplateRows :: GridTrackList a -> Css
+gridTemplateRows = key "grid-template-rows"
+
+-- | Property defines the line names and track sizing functions of the grid columns.
+gridTemplateColumns :: GridTrackList a -> Css
+gridTemplateColumns = key "grid-template-columns"
+
+newtype GridTrackList a = GridTrackList Value
+  deriving (Val, None, Inherit, Initial, Unset)
```
Comment (turion) [inline src/Clay/Grid.hs:96]:
```suggestion
  deriving (Val, None, Inherit, Initial, Unset, MaxContent, MinContent)
```

File: src/Clay/Grid.hs [hunk 25]
```diff
@@ -1,58 +1,249 @@
...
+
+-- | Property defines the line names and track sizing functions of the grid rows.
+gridTemplateRows :: GridTrackList a -> Css
+gridTemplateRows = key "grid-template-rows"
+
+-- | Property defines the line names and track sizing functions of the grid columns.
+gridTemplateColumns :: GridTrackList a -> Css
+gridTemplateColumns = key "grid-template-columns"
+
+newtype GridTrackList a = GridTrackList Value
```
Comment (turion) [inline src/Clay/Grid.hs:95]: Since `Size` is also an instance of `None`, `Initial`, etc., why not define it directly as a list of sizes?
```suggestion
newtype GridTrackList a = GridTrackList [Size a]
```

Comment (turion) [inline src/Clay/Grid.hs:95]: My understanding now is that `GridTrackList` is a concept that has no corresponding CSS name, and it mirrors the possible values one can give to `grid-template-columns` and `grid-template-rows`. Maybe the naming can reflect this, and stay closer to known concepts? So for example `GridTemplateSizes`?

File: src/Clay/Grid.hs [hunk 26]
```diff
@@ -1,58 +1,249 @@
...
+  deriving (Val, None, Inherit, Initial, Unset)
+
+-- | Create a GridTrackList from a list of sizes
+mkGridTrackList :: [Size a] -> GridTrackList a
+mkGridTrackList = GridTrackList . noCommas
```
Comment (turion) [inline src/Clay/Grid.hs:100]:
```suggestion
gridTrackList :: [Size a] -> GridTrackList a
gridTrackList = GridTrackList . noCommas
```
But see https://github.com/sebastiaanvisser/clay/pull/210/files#r780772464 first.

File: src/Clay/Grid.hs [hunk 27]
```diff
@@ -1,58 +1,249 @@
...
+gridAutoRows = key "grid-auto-rows"
 
 -- | Property defines the line names and track sizing functions of the grid columns.
-gridTemplateColumns :: [Size a] -> Css
-gridTemplateColumns = key "grid-template-columns" . noCommas
+gridAutoColumns :: GridAutoTrackList a -> Css
+gridAutoColumns = key "grid-auto-columns"
+
+newtype GridAutoTrackList a = GridAutoTrackList Value
+  deriving (Val, Auto, MinContent, MaxContent, Inherit, Initial, Unset)
```
Comment (turion) [inline src/Clay/Grid.hs:113]:
```suggestion
  deriving (Val, Auto, MinContent, MaxContent)
```
I think the others are not valid, see https://www.w3schools.com/cssref/pr_grid-auto-columns.asp

File: src/Clay/Grid.hs [hunk 28]
```diff
@@ -1,58 +1,249 @@
...
+gridAutoRows :: GridAutoTrackList a -> Css
+gridAutoRows = key "grid-auto-rows"
 
 -- | Property defines the line names and track sizing functions of the grid columns.
-gridTemplateColumns :: [Size a] -> Css
-gridTemplateColumns = key "grid-template-columns" . noCommas
+gridAutoColumns :: GridAutoTrackList a -> Css
+gridAutoColumns = key "grid-auto-columns"
+
+newtype GridAutoTrackList a = GridAutoTrackList Value
```
Comment (turion) [inline src/Clay/Grid.hs:112]: The same comments as for `GridTrackList` apply.

File: src/Clay/Grid.hs (hunk 25 above)
Comment (turion) [reply src/Clay/Grid.hs:95]: Or maybe these are called "tracks" in the community, and I just don't know?

**Review by turion — CHANGES_REQUESTED** [review]
First superficial round done. I like where this is going!

File: spec/Clay/GridSpec.hs [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+        gridTemplateRows [upcast (px 50), upcast $ fitContent (px 500), minmax (em 400) (pct 50)]
+
+  describe "gridTemplateColumns" $ do
+    describe "keyword" $ do
+      "{grid-template-columns:none}"
+        `shouldRenderFrom`
+        gridTemplateColumns none
+
+    describe "list of sizes" $ do
+      "{grid-template-columns:1em calc(20% + 1fr) auto}"
```
Comment (nsaunders) [inline spec/Clay/GridSpec.hs:46]: From the [grid spec](https://www.w3.org/TR/css-grid-2/#typedef-flex):

> Note: [&lt;flex&gt;](https://www.w3.org/TR/css-grid-2/#typedef-flex) values are not [&lt;length&gt;](https://www.w3.org/TR/css-values-3/#length-value)s (nor are they compatible with &lt;length&gt;s, like some [&lt;percentage&gt;](https://www.w3.org/TR/css-values-3/#percentage-value) values), so they cannot be represented in or combined with other unit types in [calc()](https://www.w3.org/TR/css-values-4/#funcdef-calc) expressions.

**Comment by CharlesTaylor7** [comment]
So I obviously lost steam to work on this, and there's a newer pull request in place, so withdrawing this PR. https://github.com/sebastiaanvisser/clay/pull/250
